_posts/2020-04-13-deprecation-warnings-rails-guide.markdown

```diff
@@ -0,0 +1,107 @@
+Features are deprecated rather than immediately removed, to provide backward compatibility (a solution that works in both the current and the future version), and to give programmers time to implement the code in a way that follows the new standard.
+
+In this guide we'll show you what's the workflow that we use at [FastRuby.io](https://www.fastruby.io) to address deprecation warnings when we [upgrade Rails applications](https://www.fastruby.io/blog/tags/upgrades).
+
+<!--more-->
+
+## Finding
+
+There are a few different ways to find deprecation warnings in your application.
+
+If you have good test coverage, you can run the whole test suite and look at the logs that were genarated. If you are using a CI service (like [CircleCI](https://circleci.com/) or [Travis CI](https://travis-ci.org/)) you can easily see the logs once the build finished running. Otherwise, if you run the tests locally, you can look at the output on the console or in the `log/test.log` file.
+
+All deprecation warnings start with `DEPRECATION WARNING:`, so you can search for that in the logs.
+
+When it comes to production, the easier way to collect deprecation warnings is by using a monitoring tool (like [Honeybadger](https://www.honeybadger.io/) or [Airbrake](https://airbrake.io/)).
+
+In order to send the deprecation warnings to one of these tools, you would have to do something like this:
+
+```ruby
+# config/environments/production.rb
+
+config.active_support.deprecation = :notify
+```
+
+```ruby
+# config/initializers/deprecation_warnings.rb
+
+ActiveSupport::Notifications.subscribe('deprecation.rails') do |name, start, finish, id, payload|
+  # Example if you use Honeybadger:
+  Honeybadger.notify(
+    error_class:   "deprecation_warning",
+    error_message: payload[:message],
+    backtrace:     payload[:callstack]
+  )
+end
+```
+
+See [this link](https://guides.rubyonrails.org/active_support_instrumentation.html#subscribing-to-an-event) for more details about `ActiveSupport::Notifications.subscribe`.
+
+You can also set `config.active_support.deprecation` to `:log` and look at the `log/production.log` file, but it won't be as straightforward as the first option.
+
+## Tracking
+
+Once you have all the deprecation warnings (or most of them) from your application, it is a good a idea to track them as if they were issues.
+You can use the project management tool of your preference (we use [Pivotal Tracker](https://www.ombulabs.com/blog/agile/pivotal-tracker/how-we-use-pivotal-tracker-at-ombu-labs.html)) and create a ticket in the backlog for each of the deprecation warnings. That way it makes things a lot easier when it comes to code review and organization in general.
+
+<img src="/assets/images/deprecation-warning-story.png" alt="Deprecation Warning Story for Rails Upgrade" />
```

Comment:
@lubc You might want to specify that you want to create one story per root cause, not one per deprecation warning?